torch_geometric/transforms/add_positional_encoding.py

```diff
@@ -0,0 +1,119 @@
+    r"""Adds positional encoding to the given graph (functional name:
+    :obj:`add_positional_encoding`).
+
+    Args:
+        name: (str): The type of the positional encoding
+            (:obj:`"laplacian_eigenvector_pe"`, :obj:`"random_walk_pe"`).
+            If set to :obj:`"laplacian_eigenvector_pe"`, will add the
+            Laplacian eigenvector positional encoding from the `"Benchmarking
+            Graph Neural Networks" <https://arxiv.org/abs/2003.00982>`_ paper.
+            If set to :obj:`"random_walk_pe"`, will add the random walk
+            positional encoding from the `"Graph Neural Networks with Learnable
+            Structural and Positional Representations"
+            <https://arxiv.org/abs/2110.07875>`_ paper.
+        num_channels: (int): The number of channels. If :attr:`name` is
+            :obj:`"laplacian_eigenvector_pe"`, it will be the number of the
+            smallest non-trivial eigenvectors. If :attr:`name` is
+            :obj:`"random_walk_pe"`, it will be the number of the random walk
+            steps.
+        method: (str, optional): The method to add positional encoding.
+            If set to :obj:`"attr"`, will be added to the data object
+                as a new attribute.
+            If set to :obj:`"cat"`, will be concatenated to the existing
+                :obj:`x` in the data object. (default: :obj:`attr`)
+        is_undirected (bool, optional): If set to :obj:`True`, this transform
+            expects undirected graphs as input, and can hence speed up the
+            computation of the eigenvectors. (default: :obj:`False`)
+    """
+    def __init__(self, name: str, num_channels: int,
+                 method: Optional[str] = 'attr',
+                 is_undirected: Optional[bool] = False):
+        self.name = name
+        self.num_channels = num_channels
+        self.method = method
+        self.is_undirected = is_undirected
+        assert name in ['laplacian_eigenvector_pe', 'random_walk_pe']
+        assert method in ['attr', 'cat']
+
+    def __call__(self, data: Data):
+        pe = None
+        N = data.num_nodes
+        if self.name == 'laplacian_eigenvector_pe':
+            eig_fn = eigs if not self.is_undirected else eigsh
+            edge_index, edge_weight = get_laplacian(data.edge_index,
+                                                    normalization='sym',
+                                                    dtype=torch.float,
+                                                    num_nodes=N)
+            L = to_scipy_sparse_matrix(edge_index, edge_weight, N)
+            eig_vals, eig_vecs = eig_fn(L, k=self.num_channels + 1, which='SR',
+                                        return_eigenvectors=True)
+            eig_vecs = np.real(eig_vecs[:, eig_vals.argsort()])
+            pe = torch.from_numpy(eig_vecs[:, 1:self.num_channels + 1])
+
+            sign = -1. + 2 * torch.randint(0, 2, (self.num_channels, ),
```

Comment:
That's because either positive or negative eigenvectors are both valid. To resolve this sign ambiguity, the original authors propose a random sign flip during training steps. Line 74 is implementing this procedure (multiplying -1 or 1 to eigenvectors). See page 7 in https://arxiv.org/abs/2003.00982 for more details.

Reply:
ahh okay interesting - thanks for explaining and the reference